src/android/java/io/jxcore/node/BtConnectorHelper.java

```diff
@@ -101,13 +97,15 @@ public boolean Disconnect(String peerId){
 // we only cut off our outgoing connections, incoming ones are cut off from the other end.
 // if we want to cut off whole communications, we'll do Stop
         if(mBtToRequestSocket != null) {
-            String currentpeerId = mBtToRequestSocket.GetPeerId();
-            print_debug("Disconnect : " + peerId + ", current request : " + currentpeerId);
-            if(peerId.length() == 0 || peerId.equalsIgnoreCase(currentpeerId)) {
-                print_debug("Disconnect:::Stop :" + currentpeerId);
-                mBtToRequestSocket.Stop();
-                mBtToRequestSocket = null;
-                ret = true;
+            synchronized(mBtToRequestSocket) {
+                String currentpeerId = mBtToRequestSocket.GetPeerId();
+                print_debug("Disconnect : " + peerId + ", current request : " + currentpeerId);
+                if (peerId.length() == 0 || peerId.equalsIgnoreCase(currentpeerId)) {
+                    print_debug("Disconnect:::Stop :" + currentpeerId);
+                    mBtToRequestSocket.Stop();
+                    mBtToRequestSocket = null;
```

Comment:
I would suggest having a mBtToRequestSocket class that contains all the peers we are connected to and contains a method 'disconnectPeer(peerId)' that will use a concurrent safe list with an atomic update and in that atomic update you would then check to see if the peerid exists, call close on the mBtToRequestSocket and then return null for the peer ID's state. I suspect this will be the most robust implementation that doesn't require a bunch of fiddly synchronized calls.